File: part_1/swift_solution/Tests/ExchangeLibTests/OrderTests.swift
import XCTest
@testable import ExchangeLib

class OrderTests : XCTestCase
{
    func testStringInit() throws
    {
        let order = try Order(fromString: "A:AUDUSD:100:1.47")

        XCTAssertEqual(order.participant, "A")
        XCTAssertEqual(order.instrument, "AUDUSD")
        XCTAssertEqual(order.quantity, 100)
        XCTAssertEqual(order.price, 1.47)
    }

    func testEmptyInit() throws
    {
        XCTAssertThrowsError(try Order(fromString: ""))
    }

    func testInvalidInit() throws
    {
        XCTAssertThrowsError(try Order(fromString: "blah"))
        XCTAssertThrowsError(try Order(fromString: "A:B:C:D"))
        XCTAssertThrowsError(try Order(fromString: "A:B:1.2345:1.0"))
    }

    func testWhitespaceInit() throws
    {
        let order = try Order(fromString: "  A : AUDUSD   : 100  : 1.47   ")

        XCTAssertEqual(order.participant, "A")
        XCTAssertEqual(order.instrument, "AUDUSD")
        XCTAssertEqual(order.quantity, 100)
        XCTAssertEqual(order.price, 1.47)
    }

    static var allTests =
    [
        ("testStringInit", testStringInit),
        ("testEmptyInit", testEmptyInit),
        ("testInvalidInit", testInvalidInit),
        ("testWhitespaceInit", testWhitespaceInit),
    ]
}

File: Part 1/cpp_solution/order_tests.cpp
#define BOOST_TEST_DYN_LINK
#define BOOST_TEST_MODULE order_tests
#include <boost/test/unit_test.hpp>
#include "order.h"

BOOST_AUTO_TEST_CASE( test_parse_missing_price )
{
	std::istringstream is("A:AUDUSD:100");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_invalid_price )
{
	std::istringstream is("A:AUDUSD:100:blah");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_price_rejected )
{
	std::istringstream is("A:AUDUSD:100:0");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_invalid_quantity )
{
	std::istringstream is("A:AUDUSD:xx:23.45");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_quantity )
{
	std::istringstream is("A:AUDUSD:0:23.45");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_zero_price )
{
	std::istringstream is("A:AUDUSD:1000:0");
	ae::order order;
	BOOST_REQUIRE_THROW(is >> order, std::runtime_error);
}

BOOST_AUTO_TEST_CASE( test_parse_buy )
{
	std::istringstream is("A:AUDUSD:100:1.47");
	ae::order order;
	BOOST_REQUIRE(is >> order);
    BOOST_REQUIRE_EQUAL("A", order.participant());
    BOOST_REQUIRE_EQUAL("AUDUSD", order.instrument());
    BOOST_REQUIRE_EQUAL(100, order.quantity());
    BOOST_REQUIRE_EQUAL(1.47, order.price());
    BOOST_REQUIRE(order.is_buy());
}

BOOST_AUTO_TEST_CASE( test_parse_sell )
{
	std::istringstream is("A:AUDUSD:-100:1.47");
	ae::order order;
	BOOST_REQUIRE(is >> order);
    BOOST_REQUIRE_EQUAL("A", order.participant());
    BOOST_REQUIRE_EQUAL("AUDUSD", order.instrument());
    BOOST_REQUIRE_EQUAL(100, order.quantity());
    BOOST_REQUIRE_EQUAL(1.47, order.price());
    BOOST_REQUIRE(!order.is_buy());
}

